debian/sqlite2wlk.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH SQLITE2WLK 1 "November 19, 2009"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
sqlite2wlk \- program to convert wview SQLite3 archive records to the Weatherlink WLK file format 
.SH SYNOPSIS
.B sqlite2wlk
.RI source_directory dest_directory
.br
.SH DESCRIPTION
This manual page describes the use of
.B sqlite2wlk
.
Convert wview archive data in source_directory/wview-archive.sdb to Weatherlink WLK files in 
dest_directory.
.BR Note: If Weatherlink WLK files exist in dest_directory, they will be overwritten.
.BR If source_directory is the live wview archive, you will need root access.
.SH OPTIONS
There are no options and source_directory and dest_directory are mandatory.
.TP
.SH SEE ALSO
.BR wlk2sqlite (1),
.BR /usr/share/doc/wview/wview-User-Manual.html
.br
.SH AUTHOR
sqlite2wlk was written by Mark Teel.
.PP
This manual page was written by Mark Teel <mteel2005@gmail.com>,
for the Debian project (and may be used by others).


ChangeLog
RELEASE VERSION HISTORY (newest to oldest)
------------------------------------------


5.21.5    04-06-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Remove debug message.




5.21.4    04-06-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Added stale pid file cleanup to the APT package start script.

2)  WMRUSB HID interface changes. Changed reader thread to blocking IO.
    Mitigated the console occasionally blasting out bogus data. Now zero
    packet loss. Fixed issue with pressure sanity check. Added sanity
    check for bogus rain accumulator values from the WMR console.




5.21.3    03-31-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Further WMRUSB interface improvements. Better packet framing.




5.21.2    03-30-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Updated debian APT postinst script to include copying the exfoliation skin
    for new installs.

2)  Enabled checksum processing for WMR stations.

3)  Changed LOOP windspeed items to floats for better precision.
    NOTE: If you use custom datafeed clients the LOOP data structure has changed.

4)  Added temp only packet support for WMRUSB stations. This adds support for
    add-on sensors such as the pool sensor.

5)  Fixed APT installs changing the wview.log file ownership incorrectly. Also
    restarts rsyslog automatically if a new install.




5.21.1    03-13-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Modified HILOW database storage so higher precision is used for cumulative
    data such as rain and ET.

2)  Fixed issue with yearly historical charts when generated at wview start
    (as opposed to at midnight). The current day was erroneously being included in
    the history data. This was most apparent with the UV and Solar Radiation
    charts.

3)  Fixed bug introduced in 5.21.0 release causing NULL wind direction to have
    360 added to it (to normalize calibration result). A value of -99640 was
    being stored/displayed.




5.21.0    03-05-2014
--------------------

radlib version required: 2.12.0 or newer

1)  WMRUSB improvements. Further reverse engineering of the undocumented protocol. Reset and
    heartbeat messages now working. Packet reception more stable. Initial packet acquisition
    faster resulting in faster startup.

2)  Modify CWOP packet header for new stricter compliance requirements of CWOP servers.

3)  Fix log message time_t bug for 64-bit platforms.

4)  Add timeout parameter for ssh rules. Added rows to config database.

5)  Updated data feed conversion logic to handle ARCHIVE_VALUE_NULL in loop data.

6)  Added flat line sensor value detection for outside temp, wind direction and wind speed.
    If there are no changes over a 4 hour period an email alert will be sent.

7)  Added Station Name, City and State to the wviewmgmt System Status page header.

8)  Added logic to SSH timer to fixup next timer interval if system time has skewed or
    if the ssh transfer took longer than 2 minutes.

9)  Fixed TE923 rain calculation to assume each tip is 1/36" instead of 1mm.

10) Fixed a bug for Vantage Pro stations which could cause the last archive interval of
    the previous day to be included in the current day HILOW data. This occurred if the
    midnight archive record was provided late by the console.

11) Added wet bulb temperature calculation and html tag --WetBulbTemp--. Added to all
    "readings.incx" templates.

12) Added email alert notification if no loop data has been received for a given archive
    interval. This applies to all stations except the Davis stations (alerts already
    exist for them).

13) Updated start scripts to detect and delete stale pid files.

14) Modified WH1080 data parsing to assume a humidity of 9% if the value from the
    station is undefined (0xff).




5.20.2    03-21-2012
--------------------

radlib version required: 2.12.0 or newer

1)  Added back "dailyrainin" for http reporting. They do no computation based on hourly rain values.

2)  Fixed a typo in the wviewmgmt station.php file. 




5.20.1    03-20-2012
--------------------

radlib version required: 2.12.0 or newer

1)  Modified http rain submission (Wunderground and Weatherforyou) so only the
    "rainin" parameter is sent so daily rain will be correct. 




5.20.0    03-17-2012
--------------------

radlib version required: 2.12.0 or newer

1)  Added "wait_for_time_set" function to most wview start script examples so 
    wview daemons are not started until the system time is set. This applies to
    platforms without a real time clock which must reset their time during 
    startup.

2)  Updated the User Manual section on skin selection to better describe enabling
    extended data generation (UV, ET, radiation).

3)  Updated the debian post install script to try harder to find the default http
    document root location for creation of symbolic links to wviewmgmt and the
    weather home page.

4)  Added new configuration option to optionally disable using the Vantage Pro 
    console archive records and instead autogenerate them from the LOOP data.
    This is how all other stations work. It also supports non-standard console
    interfaces (those not using the datalogger on the console).

5)  Updated the Weatherforyou URL for data submissions.

6)  Fixed bug with DataFeed clients that caused an infinite loop if you
    restarted wview while there was a datafeed client connection.

7)  Fixed bug with Alarm Type configuration introduced in 5.19.0.

8)  Changed TWI station to compute rain fall rate as it is reported to not
    be provided by the station.

9)  Added socket retry logic to the station ethernet interface (used by the
    Vantage Pro IP datalogger and any serial-to-ethernet device).

10) Modified CWOP host name to IP address resolution so that 'getaddrinfo' is
    used instead of 'gethostbyname' and all results are tried in the order they 

utilities/sqlite2wlk/sqlite2wlk.c
/*---------------------------------------------------------------------
 
 FILE NAME:
        sqlite2wlk.c
 
 PURPOSE:
        wview archive file convertor utility: SQLite3 database TO Davis WLK.
 
 REVISION HISTORY:
    Date        Programmer  Revision    Function
    09/13/2008  M.S. Teel   0           Original
 
 ASSUMPTIONS:
 None.
 
------------------------------------------------------------------------*/
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/select.h>

#include <sysdefs.h>
#include <datadefs.h>
#include <dbfiles.h>
#include "wvutilities.h"


static void USAGE (void)
{
    printf ("Usage: sqlite2wlk <source_directory> <dest_directory>\n\n");
    printf ("    Convert wview archive data in <source_directory>/wview-archive.sdb\n");
    printf ("    to Davis WLK files in <dest_directory>\n");
    printf ("Note: If Davis WLK files in <dest_directory>, they will be overwritten.\n");
    printf ("      If <source_directory> is the live wview archive, you will need root access.\n\n");
    return;
}

#define GETVALWITHOFFSET(x,y)       ((x <= ARCHIVE_VALUE_NULL) ? 0xff : (uint8_t)(x+y))
#define GETVALNOOFFSET(x)                   ((x <= ARCHIVE_VALUE_NULL) ? 0xff : (uint8_t)x)

static void convertArchivePktToWLK(ARCHIVE_PKT* archivePkt, ARCHIVE_RECORD* newRecord, time_t timeStamp)
{
    uint16_t        temp;
    time_t          tempTime;

    // create the silly packed time crap for the record:
    // First check for 24:00 madness:
    if (wvutilsGetHour(timeStamp) == 0 && wvutilsGetMin(timeStamp) == 0)
    {
        // Need to make it 24:00:
        tempTime = timeStamp - (3*WV_SECONDS_IN_HOUR);      // Avoid DST issues
        newRecord->date = INSERT_PACKED_DATE(wvutilsGetYear(tempTime),
                                             wvutilsGetMonth(tempTime),
                                             wvutilsGetDay(tempTime));
        newRecord->time = (100 * 24);
    }
    else
    {
        newRecord->date = INSERT_PACKED_DATE(wvutilsGetYear(timeStamp),
                                             wvutilsGetMonth(timeStamp),
                                             wvutilsGetDay(timeStamp));
        newRecord->time = (100 * wvutilsGetHour(timeStamp)) + wvutilsGetMin(timeStamp);
    }

    // Set the values we can:
    newRecord->outTemp          = (int16_t)(archivePkt->value[DATA_INDEX_outTemp]*10);
    newRecord->highOutTemp      = (int16_t)(archivePkt->value[DATA_INDEX_outTemp]*10);
    newRecord->lowOutTemp       = (int16_t)(archivePkt->value[DATA_INDEX_outTemp]*10);
    temp = (uint16_t)(archivePkt->value[DATA_INDEX_rain]*100);
    newRecord->rain             = temp;
    newRecord->highRainRate     = (uint16_t)(archivePkt->value[DATA_INDEX_rainRate]*100);
    newRecord->barometer        = (uint16_t)(archivePkt->value[DATA_INDEX_barometer]*1000);
    newRecord->radiation        = (uint16_t)(archivePkt->value[DATA_INDEX_radiation]);
    newRecord->windSamples      = 0xffff;
    newRecord->inTemp           = (int16_t)(archivePkt->value[DATA_INDEX_inTemp]*10);
    newRecord->inHumidity       = (uint8_t)archivePkt->value[DATA_INDEX_inHumidity];
    newRecord->outHumidity      = (uint8_t)archivePkt->value[DATA_INDEX_outHumidity];
    newRecord->avgWindSpeed     = (uint8_t)archivePkt->value[DATA_INDEX_windSpeed];
    newRecord->highWindSpeed    = (uint8_t)archivePkt->value[DATA_INDEX_windGust];
    newRecord->highWindDir      = (uint8_t)(archivePkt->value[DATA_INDEX_windGustDir] / 22.5);
    newRecord->prevWindDir      = (uint8_t)(archivePkt->value[DATA_INDEX_windDir] / 22.5);
    newRecord->UV               = (uint8_t)(archivePkt->value[DATA_INDEX_UV]*10);
    newRecord->ET               = (uint8_t)(archivePkt->value[DATA_INDEX_ET]*1000);
    newRecord->highRadiation    = (uint16_t)(archivePkt->value[DATA_INDEX_radiation]);
    newRecord->highUV           = (uint8_t)(archivePkt->value[DATA_INDEX_UV]*10);
    newRecord->fcstRule         = 0xff;
    newRecord->leafTemp1        = GETVALWITHOFFSET(archivePkt->value[DATA_INDEX_leafTemp1],90);
    newRecord->leafTemp2        = GETVALWITHOFFSET(archivePkt->value[DATA_INDEX_leafTemp2],90);
    newRecord->leafWet1         = GETVALNOOFFSET(archivePkt->value[DATA_INDEX_leafWet1]);
    newRecord->leafWet2         = GETVALNOOFFSET(archivePkt->value[DATA_INDEX_leafWet2]);
    newRecord->soilTemp1        = GETVALWITHOFFSET(archivePkt->value[DATA_INDEX_soilTemp1],90);
    newRecord->soilTemp2        = GETVALWITHOFFSET(archivePkt->value[DATA_INDEX_soilTemp2],90);
    newRecord->soilTemp3        = GETVALWITHOFFSET(archivePkt->value[DATA_INDEX_soilTemp3],90);
    newRecord->soilTemp4        = GETVALWITHOFFSET(archivePkt->value[DATA_INDEX_soilTemp4],90);
    newRecord->recordType       = 0;
    newRecord->extraHumid1      = GETVALNOOFFSET(archivePkt->value[DATA_INDEX_extraHumid1]);
    newRecord->extraHumid2      = GETVALNOOFFSET(archivePkt->value[DATA_INDEX_extraHumid2]);
    newRecord->extraTemp1       = GETVALWITHOFFSET(archivePkt->value[DATA_INDEX_extraTemp1],90);
    newRecord->extraTemp2       = GETVALWITHOFFSET(archivePkt->value[DATA_INDEX_extraTemp2],90);
    newRecord->extraTemp3       = GETVALWITHOFFSET(archivePkt->value[DATA_INDEX_extraTemp3],90);
    newRecord->soilMoist1       = GETVALNOOFFSET(archivePkt->value[DATA_INDEX_soilMoist1]);
    newRecord->soilMoist2       = GETVALNOOFFSET(archivePkt->value[DATA_INDEX_soilMoist2]);
    newRecord->soilMoist3       = GETVALNOOFFSET(archivePkt->value[DATA_INDEX_soilMoist3]);
    newRecord->soilMoist4       = GETVALNOOFFSET(archivePkt->value[DATA_INDEX_soilMoist4]);

    return;
}

// Convert SQLite3 database to WLK database:
static void ConvertSqliteToWlk (char *srcDir, char *destDir)
{
    time_t              timeStamp, diffTime, retVal, startTime = time(NULL);
    struct tm           buildTime;
    ARCHIVE_RECORD      wlkPkt;
    ARCHIVE_PKT         sqlitePkt;
    int                 inserts = 0, errors = 0;
    char                cmnd[_MAX_PATH];

    dbsqliteArchiveSetPath(srcDir);

    // ... Initialize the archive database interface:
    if (dbsqliteArchiveInit() == ERROR)
    {
        printf("dbsqliteArchiveInit failed");
        return;
    }

    sprintf(cmnd, "rm -rf %s", destDir);
    if (system(cmnd) != 0)
    {
        printf("system error:%s:%s\n", cmnd, strerror(errno));
    }
    if (mkdir(destDir, S_IFDIR | 00777) != 0)
    {
        printf("mkdir:%s:%s\n", destDir, strerror(errno));
    }

    // Set the start date to 1/1/2000:
    // Build the sqlite timestamp:
    memset (&buildTime, 0, sizeof(buildTime));
    buildTime.tm_min    = 0;
    buildTime.tm_hour   = 0;
    buildTime.tm_mday   = 1;
    buildTime.tm_mon    = 0;
    buildTime.tm_year   = 100;
    buildTime.tm_isdst  = -1;
    timeStamp = mktime(&buildTime);

    for ((retVal = dbsqliteArchiveGetNextRecord(timeStamp, &sqlitePkt)); 
         (int)retVal != ERROR;
         (retVal = dbsqliteArchiveGetNextRecord(retVal, &sqlitePkt)))
    {
        // OK, need to insert this puppy:
        convertArchivePktToWLK(&sqlitePkt, &wlkPkt, retVal);

        if (dbfStoreArchiveRecord(destDir, &wlkPkt, sqlitePkt.interval, 0x1000) == ERROR)
        {
            // Error:
            errors ++;
        }
        else
        {
            // Good stuff:
            inserts ++;
        }
    }

    diffTime = time(NULL) - startTime;

    dbsqliteArchiveExit();

    // Output results:
    printf("Conversion Stats:\n");
    printf("    Time      : %d:%d\n", (int)diffTime/60, (int)diffTime%60);
    printf("    Recs/sec  : %d\n", (inserts+errors)/(int)diffTime);
    printf("    Inserts   : %d\n", inserts);
    printf("    Errors    : %d\n", errors);
}

int main (int argc, char *argv[])
{
    char            *SourceDir, *DestDir;
    struct stat     fileData;
    
    if (argc < 3)
    {
        USAGE ();
        return ERROR;
    }

    SourceDir = argv[1];
    DestDir = argv[2];

    // sanity check the arguments
    if (stat(SourceDir, &fileData) != 0)
    {
        printf ("Source directory %s does not exist!\n", SourceDir);
        return ERROR;
    }
    else if (!(fileData.st_mode & S_IFDIR))
    {
        printf ("Source directory %s is not a directory!\n", SourceDir);
        return ERROR;
    }
    if (stat (DestDir, &fileData) != 0)
    {
        printf ("Destination directory %s does not exist, creating...\n", DestDir);
        if (mkdir(DestDir, 0755) != 0)
        {
            printf ("Destination directory %s could not be created: %s\n", 
                    DestDir, strerror(errno));
            return -1;
        }
    }

    // OK, args appear to be good, convert SQLite DB to WLK files:
    printf("Converting...(this takes ~25 minutes per year of data at 5 minute archive interval)...\n");
    ConvertSqliteToWlk (SourceDir, DestDir);

    exit (0);
}
